src/lib/cfa/nsswitch.rb

```diff
@@ -0,0 +1,260 @@
+        @file_handler.write(write_path, content)
+        @current_content = content
+      end
+    end
+
+  private
+
+    # Return the CFA::Matcher for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [CFA::Matcher]
+    def database_matcher(db_name)
+      CFA::Matcher.new { |k, v| k == "database[]" && v.value == db_name }
+    end
+
+    # Find or create the DatabaseEntry for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [DatabaseEntry]
+    def entry_for(db_name)
+      return databases[db_name] if databases.keys.include?(db_name)
+
+      database = CFA::AugeasTreeValue.new(CFA::AugeasTree.new, db_name)
+      data.add("database[]", database)
+      DatabaseEntry.new(database)
+    end
+
+    # Available databases
+    #
+    # @return [Hash{String => DatabaseEntry}]
+    def databases
+      raw_databases.each_with_object({}) do |database, collection|
+        name = database[:value].value.to_s
+        collection[name] = DatabaseEntry.new(database[:value])
+      end
+    end
+
+    # Available databases in internal structure
+    #
+    # @return [Array<Hash{Symbol => CFA::AugeasTreeValue, String, Symbol}>]
+    def raw_databases
+      data.select(CFA::Matcher.new(collection: "database"))
+    end
+
+    # Known collection keys
+    COLLECTION_KEYS = ["database", "service"].freeze
+    private_constant :COLLECTION_KEYS
+
+    # Fix collection names adding the missing "[]"
+    #
+    # When a collection has only one element, Augeas does not add "[]" suffix, reason why is
+    # necessary to fix at least those collections that are going to be hitted for either, read or
+    # write.
+    #
+    # @param tree [AugeasTree,AugeasTreeValue]
+    def fix_collection_names(tree)
+      tree.data.each do |entry|
+        entry[:key] << "[]" if COLLECTION_KEYS.include?(entry[:key])
+        fix_collection_names(entry[:value].tree) if entry[:value].is_a?(AugeasTreeValue)
+        fix_collection_names(entry[:value]) if entry[:value].is_a?(AugeasTree)
+      end
+    end
+
+    # Internal class for managing database entries
+    class DatabaseEntry
+      # Constructor
+      #
+      # @param data [CFA::AugeasTreeValue]
+      def initialize(data)
+        @data = data
+      end
+
+      # The database name
+      #
+      # @return [String]
+      def name
+        data.value
+      end
+
+      # The database service specifications
+      #
+      # @return [Array<String>]
+      def services
+        data.tree.collection("service").map(&:to_s)
+      end
+
+      # Set service specifications
+      #
+      # TODO: add support for reactions
+      #
+      # @param services [Array<String>]
+      def services=(services)
+        data.tree.delete("service[]")
+        data.tree.delete("reaction[]")
+
+        services.each { |service| data.tree.add("service[]", service) }
+      end
```

Comment:
Just for the record, a _working prototype_ for adding support to service reactions/actions


```ruby
...

# @param services [Array<Hash{Symbol => String,Hash}>] a map of services and its reactions (if any)
def services=(services)
  data.tree.delete("service[]")
  data.tree.delete("reaction[]")
 
   services.each do |service|
     data.tree.add("service[]", service[:name])
     service_reaction = service[:reaction]
   

     if (service_reaction)
        reaction = CFA::AugeasTree.new
        action = CFA::AugeasTreeValue.new(CFA::AugeasTree.new, service_reaction[:status])
        action.tree.add("action", service_reaction[:action])
        reaction.add("status", action)
        data.tree.add("reaction[]", reaction)
      }
   end
end

...
...

# $> hosts.services = [{ name: "nis", reaction: { status: "NOTFOUND", action: "return" }, { name: "files" }]
# hosts: nis [NOTFOUND=return] files
```

For sure, a final version should be much better and probably to define more precisely how the services param must be.